text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+
+The most promising operation for recovering space via pruning is the replacement of a leaf node of substantial size, which will free up the memory used by the previous incarnation of said leaf. Unfortunately the largest leaf nodes in the trie store are typically WASM blobs, thus state pruning would offer the largest size savings if contracts were frequently replaced by newer versions, which is not likely (or at least not desirable) to happen.
+
+The current implementation of the trie store has a small wrinkle in that when executing a set of operations like applying all deploy operations of a block, it creates and stores intermediate root hashes that will soon be unreachable. This adds to the set of nodes that can be pruned. If this detail were fixed, unless there is an increase in churn of WASM code, the savings from trie-store pruning are potentially not substantial.
+
+These considerations aside, pruning for the trie store can be implemented using a simple `O(n)` (assuming a hash set) garbage collection algorithm:
+
+1. Decide on a set of root hashes to keep.
+1. Traverse the tree of each root hash, adding each to a set of "nodes to keep".
+1. Iterate over the backing store, deleting all keys not in the "nodes to keep" set.
+
+A potentially more invasive mark-and-sweep version of this algorithm allows for online-collection by adding a `keep` flag to each node that defaults to `true` for every value inserted into the backing store and using the following algorithm:
+
+1. Set all `keep` flags to `false` for all currently known keys.
+1. Decide on a set of root hashes to keep.
+1. Traverse the tree of each root hash, setting `keep` to `true` for every visited key.
+1. Iterate over the backing store, deleting all keys not marked `keep`.
+
+This algorithm has the advantage of using less memory and being suspendable.
+
+This leaves us with the problem of deciding which root hashes to keep, which will be addressed later.
+
+### Storage pruning
+
+Pruning of the storage offers more substantial savings than the trie store, as most disk space is to consumed by deploys. Each block has associated data (deploys, finality signatures, etc.) that is unlikely to be shared by other blocks, as deploys are only added to one block and finality signatures are also unique. Evicting blocks from the store should thus free up space proportionally on average.
+
+By deciding to "cut" the block chain at a specific parent, we can apply a similar garbage collection algorithm:
+
+1. Decide which blocks to keep and add them to the "objects to keep" set (*i.e.* a starting block and all descendents).
```

Comment:
I don't want to rely on indices for general garbage collection to work. The "objects to keep" set also has an advantage, it makes it possible to garbage collect a running system: When the GC is running, every newly stored object is added to the "objects to keep" set when inserted and thus not touched.

Or put differently, like specifying what to keep and removing the rest instead of the opposite, since it clears out any strays :)

Reply:
What is preventing us from running GC in the background @Fraser999's solution?  

One drawback of the keep list based solution is that it would need to be persisted to disk in the event that the node is stopped and restarted.